{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.special as sc\n",
    "import matplotlib as m\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# control experiment\n",
    "\n",
    "# 3 reps, 100 iterations \n",
    "# samples from ONLY WT!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all bayexpress results\n",
    "# 3 replicates\n",
    "# BF > 100\n",
    "\n",
    "CONTROL_R3_bayexpress_FC0 = pd.read_csv(f'DGE_results/CONTROL_R3_1_bayexpress.csv').iloc[:,1:2]\n",
    "CONTROL_R3_bayexpress_FC1 = pd.read_csv(f'DGE_results/CONTROL_R3_1_bayexpress.csv').iloc[:,1:2]\n",
    "CONTROL_R3_bayexpress_FC2 = pd.read_csv(f'DGE_results/CONTROL_R3_1_bayexpress.csv').iloc[:,1:2]\n",
    "\n",
    "for i in range(1,101):\n",
    "\n",
    "    data = pd.read_csv(f'DGE_results/CONTROL_R3_{i}_bayexpress.csv').iloc[:,1:]\n",
    "\n",
    "    # criteria 0 for significance: (BF > 100) & (abs(FC) > 0)\n",
    "    CONTROL_R3_bayexpress_FC0[f'{i}'] = (data.BF > 100) & (abs(data.FC) > 0)\n",
    "\n",
    "    # criteria 1 for significance: (BF > 100) & (abs(FC) > 1)\n",
    "    CONTROL_R3_bayexpress_FC1[f'{i}'] = (data.BF > 100) & (abs(data.FC) > 1)\n",
    "\n",
    "    # criteria 2 for significance: (BF > 100) & (abs(FC) > 2)\n",
    "    CONTROL_R3_bayexpress_FC2[f'{i}'] = (data.BF > 100) & (abs(data.FC) > 2)\n",
    "\n",
    "CONTROL_R3_bayexpress_FC0['Ntrue'] = [(CONTROL_R3_bayexpress_FC0.iloc[i].value_counts().get(True, 0)) for i in range(len(CONTROL_R3_bayexpress_FC0))]\n",
    "\n",
    "CONTROL_R3_bayexpress_FC1['Ntrue'] = [(CONTROL_R3_bayexpress_FC1.iloc[i].value_counts().get(True, 0)) for i in range(len(CONTROL_R3_bayexpress_FC1))]\n",
    "\n",
    "CONTROL_R3_bayexpress_FC2['Ntrue'] = [(CONTROL_R3_bayexpress_FC2.iloc[i].value_counts().get(True, 0)) for i in range(len(CONTROL_R3_bayexpress_FC2))]\n",
    "\n",
    "display(CONTROL_R3_bayexpress_FC0)\n",
    "display(CONTROL_R3_bayexpress_FC1)\n",
    "display(CONTROL_R3_bayexpress_FC2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1, 1, sharey=True, figsize=(6, 4), dpi=300)\n",
    "\n",
    "# scatters\n",
    "axs.scatter(np.arange(1, 101),\n",
    "            CONTROL_R3_bayexpress_FC0.iloc[:, 1:-1].sum(axis=0),\n",
    "            c='#332288', s=30,\n",
    "            label=f'BF > 100, FC > 0, total = {sum(CONTROL_R3_bayexpress_FC0.iloc[:,1:-1].sum(axis=0))}',\n",
    "            alpha=0.9, marker='x')\n",
    "\n",
    "axs.scatter(np.arange(1, 101),\n",
    "            CONTROL_R3_bayexpress_FC1.iloc[:, 1:-1].sum(axis=0),\n",
    "            c='#44AA99', s=30,\n",
    "            label=f'BF > 100, FC > 1, total = {sum(CONTROL_R3_bayexpress_FC1.iloc[:,1:-1].sum(axis=0))}',\n",
    "            alpha=0.9, marker='x')\n",
    "\n",
    "axs.scatter(np.arange(1, 101),\n",
    "            CONTROL_R3_bayexpress_FC2.iloc[:, 1:-1].sum(axis=0),\n",
    "            c='#CC6677', s=30,\n",
    "            label=f'BF > 100, FC > 2, total = {sum(CONTROL_R3_bayexpress_FC2.iloc[:,1:-1].sum(axis=0))}',\n",
    "            alpha=0.9, marker='x')\n",
    "\n",
    "axs.legend(loc='upper left')\n",
    "\n",
    "# Use tex in labels\n",
    "axs.set_xticks(np.arange(1, 101))\n",
    "axs.set_xticklabels('')\n",
    "\n",
    "# formatting y axis ticks\n",
    "plt.gca().ticklabel_format(axis='y', style='plain', useOffset=False)\n",
    "\n",
    "axs.set_xlabel('100 Bootstrapping iterations')\n",
    "axs.set_ylabel('# putative differentially expressed genes')\n",
    "\n",
    "# Set y limit 3500\n",
    "axs.set_ylim(-100, 3500)\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# zoom in\n",
    "\n",
    "fig, axs = plt.subplots(1, 1, sharey=True, figsize=(6, 4), dpi=300)\n",
    "\n",
    "# scatters\n",
    "axs.scatter(np.arange(1, 101),\n",
    "            CONTROL_R3_bayexpress_FC0.iloc[:, 1:-1].sum(axis=0),\n",
    "            c='#332288', s=30,\n",
    "            label=f'BF > 100, FC > 0, total = {sum(CONTROL_R3_bayexpress_FC0.iloc[:,1:-1].sum(axis=0))}',\n",
    "            alpha=0.9, marker='x')\n",
    "\n",
    "axs.scatter(np.arange(1, 101),\n",
    "            CONTROL_R3_bayexpress_FC1.iloc[:, 1:-1].sum(axis=0),\n",
    "            c='#44AA99', s=30,\n",
    "            label=f'BF > 100, FC > 1, total = {sum(CONTROL_R3_bayexpress_FC1.iloc[:,1:-1].sum(axis=0))}',\n",
    "            alpha=0.9, marker='x')\n",
    "\n",
    "axs.scatter(np.arange(1, 101),\n",
    "            CONTROL_R3_bayexpress_FC2.iloc[:, 1:-1].sum(axis=0),\n",
    "            c='#CC6677', s=30,\n",
    "            label=f'BF > 100, FC > 2, total = {sum(CONTROL_R3_bayexpress_FC2.iloc[:,1:-1].sum(axis=0))}',\n",
    "            alpha=0.9, marker='x')\n",
    "\n",
    "axs.legend(loc='upper left')\n",
    "\n",
    "# Use tex in labels\n",
    "axs.set_xticks(np.arange(1, 101))\n",
    "axs.set_xticklabels('')\n",
    "\n",
    "# formatting y axis ticks\n",
    "plt.gca().ticklabel_format(axis='y', style='plain', useOffset=False)\n",
    "\n",
    "axs.set_xlabel('100 Bootstrapping iterations')\n",
    "axs.set_ylabel('# putative differentially expressed genes')\n",
    "\n",
    "# Set y limit to 300\n",
    "axs.set_ylim(-10, 300)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# just a different visualisation\n",
    "\n",
    "\n",
    "fig, axs = plt.subplots(1,1,sharey=True, figsize=(6,4), dpi=300)\n",
    "\n",
    "\n",
    "series = CONTROL_R3_bayexpress_FC0.iloc[:,1:-1].sum(axis=0)\n",
    "series.plot(kind='line', color='#332288', alpha=0.8, ylim=[-5,500], linewidth=2, label=f'BF > 100, FC > 0, total = {sum(CONTROL_R3_bayexpress_FC0.iloc[:,1:-1].sum(axis=0))}')\n",
    "\n",
    "series = CONTROL_R3_bayexpress_FC1.iloc[:,1:-1].sum(axis=0)\n",
    "series.plot(kind='line', color='#44AA99', alpha=0.8, ylim=[-5,500], linewidth=2, label=f'BF > 100, FC > 1, total = {sum(CONTROL_R3_bayexpress_FC1.iloc[:,1:-1].sum(axis=0))}')\n",
    "\n",
    "series = CONTROL_R3_bayexpress_FC2.iloc[:,1:-1].sum(axis=0)\n",
    "series.plot(kind='line', color='#CC6677', alpha=0.8, ylim=[-5,500], linewidth=2, label=f'BF > 100, FC > 2, total = {sum(CONTROL_R3_bayexpress_FC2.iloc[:,1:-1].sum(axis=0))}')\n",
    "\n",
    "axs.set_xlabel('Bootstrapping iterations')\n",
    "axs.set_ylabel('# putative differentially expressed genes')\n",
    "\n",
    "axs.set_xticks([1,100])\n",
    "axs.set_xticklabels(['1', '100'])\n",
    "\n",
    "axs.legend(loc='center left')\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# R3 means number of replicates in 100 bootstrapping iterations is 3\n",
    "# FC > 0\n",
    "# got hits (any)\n",
    "\n",
    "# let's find out which genes we found to produce hits in the control and see if we can identify them with Bayes factors for consistency\n",
    "\n",
    "print('Find genes that got hits in the control experiment')\n",
    "display(CONTROL_R3_bayexpress_FC0.loc[CONTROL_R3_bayexpress_FC0.Ntrue > 0])\n",
    "\n",
    "# importing results from differential gene expression anaylsis to check what they are\n",
    "RALL_bayexpress = pd.read_csv('RALL_bayexpress.csv', index_col=0)\n",
    "\n",
    "print('All replicates, bayexpress results')\n",
    "display(RALL_bayexpress)\n",
    "\n",
    "print('Display only the genes that got hits in the control experiment')\n",
    "display(RALL_bayexpress.iloc[(CONTROL_R3_bayexpress_FC0.loc[CONTROL_R3_bayexpress_FC0.Ntrue > 0].index)])\n",
    "\n",
    "print('Display only the genes that got hits in the control experiment + where identified with bayes factors for consistency of replicates in the WT')\n",
    "display(RALL_bayexpress.iloc[(CONTROL_R3_bayexpress_FC0.loc[CONTROL_R3_bayexpress_FC0.Ntrue > 0].index)].loc[(RALL_bayexpress.nBF_WT > 1)])\n",
    "\n",
    "print('Display only the genes that got hits in the control experiment + where NOT identified with bayes factors for consistency of replicates in the WT')\n",
    "display(RALL_bayexpress.iloc[(CONTROL_R3_bayexpress_FC0.loc[CONTROL_R3_bayexpress_FC0.Ntrue > 0].index)].loc[(RALL_bayexpress.nBF_WT < 1)])\n",
    "\n",
    "print('Among the genes that produced 0 hits, where there any genes that would be identified as variable by nBF > 1?')\n",
    "display(RALL_bayexpress.iloc[(CONTROL_R3_bayexpress_FC0.loc[CONTROL_R3_bayexpress_FC0.Ntrue == 0].index)].loc[(RALL_bayexpress.nBF_WT > 1)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# R3 means number of replicates in 100 bootstrapping iterations is 3\n",
    "# FC > 0\n",
    "# got more than 10 hits\n",
    "\n",
    "# let's find out which genes we found to produce hits in the control and see if we can identify them with Bayes factors for consistency\n",
    "\n",
    "print('Find genes that got more than 10 hits in the control experiment')\n",
    "display(CONTROL_R3_bayexpress_FC0.loc[CONTROL_R3_bayexpress_FC0.Ntrue > 10])\n",
    "\n",
    "# importing results from differential gene expression anaylsis to check what they are\n",
    "RALL_bayexpress = pd.read_csv('RALL_bayexpress.csv', index_col=0)\n",
    "\n",
    "print('All replicates, bayexpress results')\n",
    "display(RALL_bayexpress)\n",
    "\n",
    "print('Display only the genes that got more than 10 hits in the control experiment')\n",
    "display(RALL_bayexpress.iloc[(CONTROL_R3_bayexpress_FC0.loc[CONTROL_R3_bayexpress_FC0.Ntrue > 10].index)])\n",
    "\n",
    "print('Display only the genes that got more than 10 hits in the control experiment + where identified with bayes factors for consistency of replicates in the WT')\n",
    "display(RALL_bayexpress.iloc[(CONTROL_R3_bayexpress_FC0.loc[CONTROL_R3_bayexpress_FC0.Ntrue > 10].index)].loc[(RALL_bayexpress.nBF_WT > 1)])\n",
    "\n",
    "print('Display only the genes that got more than 10 hits in the control experiment + where NOT identified with bayes factors for consistency of replicates in the WT')\n",
    "display(RALL_bayexpress.iloc[(CONTROL_R3_bayexpress_FC0.loc[CONTROL_R3_bayexpress_FC0.Ntrue > 10].index)].loc[(RALL_bayexpress.nBF_WT < 1)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# R3 means number of replicates in 100 bootstrapping iterations is 3\n",
    "# FC > 1\n",
    "# got hits (any)\n",
    "\n",
    "# let's find out which genes we found to produce hits in the control and see if we can identify them with Bayes factors for consistency\n",
    "\n",
    "print('Find genes that got hits in the control experiment')\n",
    "display(CONTROL_R3_bayexpress_FC1.loc[CONTROL_R3_bayexpress_FC1.Ntrue > 0])\n",
    "\n",
    "# importing results from differential gene expression anaylsis to check what they are\n",
    "RALL_bayexpress = pd.read_csv('RALL_bayexpress.csv', index_col=0)\n",
    "\n",
    "print('All replicates, bayexpress results')\n",
    "display(RALL_bayexpress)\n",
    "\n",
    "print('Display only the genes that got hits in the control experiment')\n",
    "display(RALL_bayexpress.iloc[(CONTROL_R3_bayexpress_FC1.loc[CONTROL_R3_bayexpress_FC1.Ntrue > 0].index)])\n",
    "\n",
    "print('Display only the genes that got hits in the control experiment + where identified with bayes factors for consistency of replicates in the WT')\n",
    "display(RALL_bayexpress.iloc[(CONTROL_R3_bayexpress_FC1.loc[CONTROL_R3_bayexpress_FC1.Ntrue > 0].index)].loc[(RALL_bayexpress.nBF_WT > 1)])\n",
    "\n",
    "print('Display only the genes that got hits in the control experiment + where NOT identified with bayes factors for consistency of replicates in the WT')\n",
    "display(RALL_bayexpress.iloc[(CONTROL_R3_bayexpress_FC1.loc[CONTROL_R3_bayexpress_FC1.Ntrue > 0].index)].loc[(RALL_bayexpress.nBF_WT < 1)])\n",
    "\n",
    "print('Among the genes that produced 0 hits, where there any genes that would be identified as variable by nBF > 1?')\n",
    "display(RALL_bayexpress.iloc[(CONTROL_R3_bayexpress_FC1.loc[CONTROL_R3_bayexpress_FC1.Ntrue == 0].index)].loc[(RALL_bayexpress.nBF_WT > 1)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.11 ('pythonbase')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "e8c6449f11eab0fe7190b9942659b44b2071ad732b0c42543511aa2c7698ac28"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
