{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as m\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing yeast count data\n",
    "\n",
    "# 42 clean WT replicates\n",
    "WT_yeast = pd.read_csv('WT_yeast.csv', index_col=0)\n",
    "\n",
    "# 44 clean Snf2 mutant replciates\n",
    "Snf2_yeast = pd.read_csv('Snf2_yeast.csv', index_col=0)\n",
    "\n",
    "display(WT_yeast)\n",
    "display(Snf2_yeast)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing yeast count data >>> q-values\n",
    "# from explore_clean_yeast_consistency.ipynb\n",
    "\n",
    "# 42 clean WT replicates\n",
    "WT_yeast_q = pd.read_csv('WT_yeast_q.csv', index_col=0)\n",
    "\n",
    "# 44 clean Snf2 mutant replciates\n",
    "Snf2_yeast_q = pd.read_csv('Snf2_yeast_q.csv', index_col=0)\n",
    "\n",
    "display(WT_yeast_q)\n",
    "display(Snf2_yeast_q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing length information from https://bluegenes.yeastgenome.org/yeastmine/results/ALL_Yeast_Genes (downloaded on 13-05-2024)\n",
    "\n",
    "lengths = pd.read_csv('yeastmine_results_2024-05-13T13-30-37.csv', usecols=['Gene > Systematic Name', 'Gene > Length', 'Gene > Transcripts > Length']).rename(columns={'Gene > Systematic Name': 'locus_name', 'Gene > Length': ' gene_length', 'Gene > Transcripts > Length': 'transcript_length'}).groupby(['locus_name'], \n",
    "as_index=False).mean()\n",
    "\n",
    "display(lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing results from differential gene expression anaylsis\n",
    "RALL_bayexpress = pd.read_csv('RALL_bayexpress.csv', index_col=0)\n",
    "\n",
    "# adding estimated q-value for each gene to the data frame\n",
    "\n",
    "RALL_bayexpress['q'] = (WT_yeast.iloc[:,1:].sum(axis=1)+1)/(sum(WT_yeast.iloc[:,1:].sum(axis=1))+2)\n",
    "\n",
    "RALL_bayexpress = RALL_bayexpress.merge(lengths, how='right', on='locus_name')\n",
    "\n",
    "\n",
    "display(RALL_bayexpress)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How does transcript length relate to Bayes factors?\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(dpi=300)\n",
    "\n",
    "ax.grid()\n",
    "\n",
    "ax.scatter(RALL_bayexpress.transcript_length, RALL_bayexpress.BF,\n",
    "            \n",
    "c='#332288', s=30, \n",
    "alpha=0.3, edgecolors='none')\n",
    "\n",
    "ax.set_xlabel('transcript length')\n",
    "ax.set_ylabel('Bayes factor')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# ------------\n",
    "\n",
    "fig, ax = plt.subplots(dpi=300)\n",
    "\n",
    "ax.grid()\n",
    "ax.set_yscale(\"log\", base=10)\n",
    "\n",
    "ax.scatter(RALL_bayexpress.transcript_length, RALL_bayexpress.BF,\n",
    "            \n",
    "c='#332288', s=30, \n",
    "alpha=0.3, edgecolors='none')\n",
    "\n",
    "ax.set_xlabel('transcript length')\n",
    "ax.set_ylabel('Bayes factor')\n",
    "\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# ------------\n",
    "\n",
    "fig, ax = plt.subplots(dpi=300)\n",
    "\n",
    "ax.grid()\n",
    "\n",
    "ax.set_xscale(\"log\", base=10)\n",
    "ax.set_yscale(\"log\", base=10)\n",
    "\n",
    "ax.scatter(RALL_bayexpress.transcript_length, RALL_bayexpress.BF,\n",
    "            \n",
    "c='#332288', s=30, \n",
    "alpha=0.3, edgecolors='none')\n",
    "\n",
    "ax.set_xlabel('transcript length')\n",
    "ax.set_ylabel('Bayes factor')\n",
    "\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How does transcript length relate to estimated q-values?\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(dpi=300)\n",
    "\n",
    "ax.grid()\n",
    "\n",
    "ax.scatter(RALL_bayexpress.transcript_length, RALL_bayexpress.q,\n",
    "            \n",
    "c='#332288', s=30, \n",
    "alpha=0.3, edgecolors='none')\n",
    "\n",
    "ax.set_xlabel('transcript length')\n",
    "ax.set_ylabel('q')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# ------------\n",
    "\n",
    "fig, ax = plt.subplots(dpi=300)\n",
    "\n",
    "ax.grid()\n",
    "ax.set_yscale(\"log\", base=10)\n",
    "\n",
    "ax.scatter(RALL_bayexpress.transcript_length, RALL_bayexpress.q,\n",
    "            \n",
    "c='#332288', s=30, \n",
    "alpha=0.3, edgecolors='none')\n",
    "\n",
    "ax.set_xlabel('transcript length')\n",
    "ax.set_ylabel('q')\n",
    "\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# ------------\n",
    "\n",
    "fig, ax = plt.subplots(dpi=300)\n",
    "\n",
    "ax.grid()\n",
    "\n",
    "ax.set_xscale(\"log\", base=10)\n",
    "ax.set_yscale(\"log\", base=10)\n",
    "\n",
    "ax.scatter(RALL_bayexpress.transcript_length, RALL_bayexpress.q,\n",
    "            \n",
    "c='#332288', s=30, \n",
    "alpha=0.3, edgecolors='none')\n",
    "\n",
    "ax.set_xlabel('transcript length')\n",
    "ax.set_ylabel('q')\n",
    "\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How does gene length relate to Bayes factors?\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(dpi=300)\n",
    "\n",
    "ax.grid()\n",
    "\n",
    "ax.scatter(RALL_bayexpress.transcript_length, RALL_bayexpress.BF,\n",
    "            \n",
    "c='#332288', s=30, \n",
    "alpha=0.3, edgecolors='none')\n",
    "\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# ------------\n",
    "\n",
    "fig, ax = plt.subplots(dpi=300)\n",
    "\n",
    "ax.grid()\n",
    "ax.set_yscale(\"log\", base=10)\n",
    "\n",
    "ax.scatter(RALL_bayexpress.transcript_length, RALL_bayexpress.BF,\n",
    "            \n",
    "c='#332288', s=30, \n",
    "alpha=0.3, edgecolors='none')\n",
    "\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# ------------\n",
    "\n",
    "fig, ax = plt.subplots(dpi=300)\n",
    "\n",
    "ax.grid()\n",
    "\n",
    "ax.set_xscale(\"log\", base=10)\n",
    "ax.set_yscale(\"log\", base=10)\n",
    "\n",
    "ax.scatter(RALL_bayexpress.transcript_length, RALL_bayexpress.BF,\n",
    "            \n",
    "c='#332288', s=30, \n",
    "alpha=0.3, edgecolors='none')\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pythonbase",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
