{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as m\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import wasserstein_distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing yeast count data\n",
    "\n",
    "# 42 clean WT replicates\n",
    "WT_yeast = pd.read_csv('WT_yeast.csv', index_col=0)\n",
    "\n",
    "# 44 clean Snf2 mutant replciates\n",
    "Snf2_yeast = pd.read_csv('Snf2_yeast.csv', index_col=0)\n",
    "\n",
    "display(WT_yeast)\n",
    "display(Snf2_yeast)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing yeast count data >>> q-values\n",
    "# from explore_clean_yeast_consistency.ipynb\n",
    "\n",
    "# 42 clean WT replicates\n",
    "WT_yeast_q = pd.read_csv('WT_yeast_q.csv', index_col=0)\n",
    "\n",
    "# 44 clean Snf2 mutant replciates\n",
    "Snf2_yeast_q = pd.read_csv('Snf2_yeast_q.csv', index_col=0)\n",
    "\n",
    "display(WT_yeast_q)\n",
    "display(Snf2_yeast_q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing yeast count data >>> average q-values\n",
    "# from explore_clean_yeast_consistency.ipynb\n",
    "\n",
    "# 42 clean WT replicates\n",
    "WT_yeast_avq = pd.read_csv('WT_yeast_avq.csv', index_col=0)\n",
    "\n",
    "# 44 clean Snf2 mutant replciates\n",
    "Snf2_yeast_avq = pd.read_csv('Snf2_yeast_avq.csv', index_col=0)\n",
    "\n",
    "display(WT_yeast_avq)\n",
    "display(Snf2_yeast_avq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing results from differential gene expression anaylsis\n",
    "RALL_bayexpress = pd.read_csv('RALL_bayexpress.csv', index_col=0)\n",
    "\n",
    "RALL_bayexpress['W'] = [wasserstein_distance(list(WT_yeast.iloc[i,1:]), list(Snf2_yeast.iloc[i,3:])) for i in range(len(RALL_bayexpress))]\n",
    "\n",
    "display(RALL_bayexpress)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing bootstrapping results\n",
    "# via package_comp_numbers.ipynb\n",
    "\n",
    "bayexpress_BFNtrue = pd.read_csv('bayexpress_BFNtrue.csv').set_index('locus_name')\n",
    "display(bayexpress_BFNtrue)\n",
    "\n",
    "bayexpress_Ntrue = pd.read_csv('bayexpress_Ntrue.csv').set_index('locus_name')\n",
    "display(bayexpress_Ntrue)\n",
    "\n",
    "edgeR_Ntrue = pd.read_csv('edgeR_Ntrue.csv').set_index('genes')\n",
    "display(edgeR_Ntrue)\n",
    "\n",
    "DESeq2_Ntrue = pd.read_csv('DESeq2_Ntrue.csv').set_index('genes')\n",
    "display(DESeq2_Ntrue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define function to get plots showing the package comparison numbers\n",
    "# all the info from explore_clean_yeast.ipynb\n",
    "# plus consistency tests\n",
    "# PCN plots, 100 bootstrapping iterations\n",
    "\n",
    "\n",
    "def stalk(genes):\n",
    "\n",
    "    # Find the index of all genes\n",
    "    igenes = [list(RALL_bayexpress.locus_name).index(gene) for gene in genes]\n",
    "    \n",
    "    # Some more numbers about the genes\n",
    "\n",
    "    print('Read counts in WT across replicates:')\n",
    "    display(WT_yeast.iloc[igenes])\n",
    "\n",
    "    print('Read counts in Snf2 mutant across replicates:')\n",
    "    display(Snf2_yeast.iloc[igenes])\n",
    "\n",
    "    print('bayexpress results:')\n",
    "    display(RALL_bayexpress.iloc[igenes])\n",
    "\n",
    "    print('With:')\n",
    "    print('BF ... Bayes factor for differential gene expression analysis')\n",
    "    print('FC ... inferred log_2 fold change')\n",
    "    print('nBF_WT ... Bayes factor for testing wether replicates are consistent with each other in wild-type')\n",
    "    print('AOTP_WT ... genes that have been identified to be All Over The Place in WT meaning it is part of the list of highy variable genes idenified in bootstrapping experiments. If AOTP == True, the gene is marked*.')\n",
    "    print('nBF_Snf2 ... Bayes factor for testing wether replicates are consistent with each other in Snf2-mutant')\n",
    "    print('AOTP_Snf2 ... genes that have been identified to be All Over The Place in the mutant meaning it is part of the list of highy variable genes idenified in bootstrapping experiments. If AOTP == True, the gene is marked*.')\n",
    "\n",
    "    # printing q-plots\n",
    "    for t in igenes:\n",
    "        fig, ax = plt.subplots(dpi=300)\n",
    "\n",
    "        ax.hist(WT_yeast_q.iloc[t], 10, density=False, histtype='step', color='#332288',\n",
    "                alpha=0.4, orientation='horizontal')\n",
    "\n",
    "        ax.hist(Snf2_yeast_q.iloc[t], 10, density=False, histtype='step', color='#DDCC77',\n",
    "                alpha=0.4, orientation='horizontal')\n",
    "\n",
    "\n",
    "        # scatter WT\n",
    "        gene_WT_nBF = str(RALL_bayexpress.loc[t].nBF_WT.round(3))\n",
    "        gene_ATOP_WT = RALL_bayexpress.loc[t].AOTP_WT\n",
    "        ax.scatter(np.arange(1,len(WT_yeast_q.columns)+1), WT_yeast_q.iloc[t],\n",
    "                \n",
    "            c='#332288', s=30, \n",
    "            label=f'BF = {gene_WT_nBF}{\"*\" if gene_ATOP_WT else \"\"}',\n",
    "            alpha=0.9, edgecolors='none')\n",
    "        \n",
    "        # scatter Snf2\n",
    "        gene_Snf2_nBF = str(RALL_bayexpress.loc[t].nBF_Snf2.round(3))\n",
    "        gene_ATOP_Snf2 = RALL_bayexpress.loc[t].AOTP_Snf2\n",
    "        ax.scatter(np.arange(1,len(Snf2_yeast_q.columns)+1), Snf2_yeast_q.iloc[t],\n",
    "                \n",
    "            c='#DDCC77', s=30, \n",
    "            label=f'BF = {gene_Snf2_nBF}{\"*\" if gene_ATOP_Snf2 else \"\"}',\n",
    "            alpha=0.9, edgecolors='none')\n",
    "        \n",
    "        ax.legend(loc='upper right')\n",
    "        \n",
    "        # average estimates WT\n",
    "        series = WT_yeast_avq.iloc[t,1:]\n",
    "        series.plot(color='#332288', alpha=0.3, linewidth=10)\n",
    "        \n",
    "        # average estimates Snf2\n",
    "        series = Snf2_yeast_avq.iloc[t,1:]\n",
    "        series.plot(color='#DDCC77', alpha=0.3, linewidth=10)\n",
    "            \n",
    "        gene = str(RALL_bayexpress.loc[t].locus_name)\n",
    "        gene_BF = str(RALL_bayexpress.loc[t].BF.round(3))\n",
    "        gene_iFC = str(RALL_bayexpress.loc[t].FC.round(3))\n",
    "        gene_W = str(RALL_bayexpress.loc[t].W.round(3))\n",
    "        ax.set_title(f'{gene}: BF = {gene_BF}, iFC = {gene_iFC}, W = {gene_W}')\n",
    "\n",
    "        # Use tex in labels\n",
    "        ax.set_xticks(np.arange(1,44))\n",
    "        ax.set_xticklabels('')\n",
    "\n",
    "        # formatting y axis ticks\n",
    "        plt.gca().ticklabel_format(axis='y', style='plain', useOffset=False)\n",
    "\n",
    "        ax.set_xlabel('Replicates')\n",
    "        ax.set_ylabel('q = (n+1) / (N+2)')\n",
    "\n",
    "        plt.show()\n",
    "\n",
    "    # package number comparison plots\n",
    "    reps = ['3/3 R', '6/6 R', '12/12 R', '20/20 R']\n",
    "\n",
    "    fig, (ax1, ax2, ax3, ax4) = plt.subplots(1,4, dpi=300, figsize=(8,3))\n",
    "\n",
    "    # Bayes factors\n",
    "\n",
    "    im = ax1.imshow(bayexpress_Ntrue.loc[genes].values, cmap='cividis', alpha=0.3, norm=m.colors.Normalize(vmin=0, vmax=100))\n",
    "\n",
    "    ax1.set_xticks(np.arange(len(reps)), labels=reps)\n",
    "    ax1.set_yticks(np.arange(len(genes)), labels=genes)\n",
    "\n",
    "    # Rotate the tick labels and set their alignment.\n",
    "    plt.setp(ax1.get_xticklabels(), rotation=45, ha=\"right\",\n",
    "            rotation_mode=\"anchor\")\n",
    "\n",
    "    # Loop over data dimensions and create text annotations.\n",
    "    for i in range(len(genes)):\n",
    "        for j in range(len(reps)):\n",
    "            text = ax1.text(j, i, bayexpress_Ntrue.loc[genes].values[i, j],\n",
    "                        ha=\"center\", va=\"center\", color=\"black\")\n",
    "    ax1.set_title('BF')\n",
    "\n",
    "    # edgeR\n",
    "\n",
    "    im = ax2.imshow(edgeR_Ntrue.loc[genes].values, cmap='cividis', alpha=0.3, norm=m.colors.Normalize(vmin=0, vmax=100))\n",
    "\n",
    "    ax2.set_xticks(np.arange(len(reps)), labels=reps)\n",
    "    ax2.set_yticks([], labels=[])\n",
    "\n",
    "    # Rotate the tick labels and set their alignment.\n",
    "    plt.setp(ax2.get_xticklabels(), rotation=45, ha=\"right\",\n",
    "            rotation_mode=\"anchor\")\n",
    "\n",
    "    # Loop over data dimensions and create text annotations.\n",
    "    for i in range(len(genes)):\n",
    "        for j in range(len(reps)):\n",
    "            text = ax2.text(j, i, edgeR_Ntrue.loc[genes].values[i, j],\n",
    "                        ha=\"center\", va=\"center\", color=\"black\")\n",
    "    ax2.set_title('edgeR')\n",
    "\n",
    "\n",
    "\n",
    "    # DESeq2\n",
    "\n",
    "    im = ax3.imshow(DESeq2_Ntrue.loc[genes].values, cmap='cividis', alpha=0.3, norm=m.colors.Normalize(vmin=0, vmax=100))\n",
    "\n",
    "    ax3.set_xticks(np.arange(len(reps)), labels=reps)\n",
    "    ax3.set_yticks([], labels=[])\n",
    "\n",
    "    # Rotate the tick labels and set their alignment.\n",
    "    plt.setp(ax3.get_xticklabels(), rotation=45, ha=\"right\",\n",
    "            rotation_mode=\"anchor\")\n",
    "\n",
    "    # Loop over data dimensions and create text annotations.\n",
    "    for i in range(len(genes)):\n",
    "        for j in range(len(reps)):\n",
    "            text = ax3.text(j, i, DESeq2_Ntrue.loc[genes].values[i, j],\n",
    "                        ha=\"center\", va=\"center\", color=\"black\")\n",
    "    ax3.set_title('DESeq2')\n",
    "\n",
    "\n",
    "    # Bayes factors only\n",
    "\n",
    "    im = ax4.imshow(bayexpress_BFNtrue.loc[genes].values, cmap='cividis', alpha=0.3, norm=m.colors.Normalize(vmin=0, vmax=100))\n",
    "\n",
    "    ax4.set_xticks(np.arange(len(reps)), labels=reps)\n",
    "    ax4.set_yticks([], labels=[])\n",
    "\n",
    "    # Rotate the tick labels and set their alignment.\n",
    "    plt.setp(ax4.get_xticklabels(), rotation=45, ha=\"right\",\n",
    "            rotation_mode=\"anchor\")\n",
    "\n",
    "    # Loop over data dimensions and create text annotations.\n",
    "    for i in range(len(genes)):\n",
    "        for j in range(len(reps)):\n",
    "            text = ax4.text(j, i, bayexpress_BFNtrue.loc[genes].values[i, j],\n",
    "                        ha=\"center\", va=\"center\", color=\"black\")\n",
    "    ax4.set_title('BF > 1')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stalk(['YAL016C-B', 'YAL031W-A', 'YAR035W', 'YAR068W'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stalk(['YGL228W', 'YBR078W', 'YIL094C', 'YGL253W'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stalk(['YGR192C', 'YOR383C', 'YHR174W', 'YDR077W'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stalk(['YNL232W', 'YLR329W', 'YDR291W', 'YPR164W'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stalk(['snR32', 'YPL032C', 'YNL034W', 'YPL030W'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.neighbors import KernelDensity\n",
    "import numpy as np\n",
    "X = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2]])\n",
    "kde = KernelDensity(kernel='gaussian', bandwidth=0.2).fit(X)\n",
    "kde.score_samples(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stalk(['YHR174W'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stalk(['YDL062W'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pythonbase",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
